doc/radosgw/s3select.rst

```diff
@@ -0,0 +1,265 @@
+    | The s3-select main flow is initiated with parsing of input-string (i.e user query), and follows 
+    | with building an AST (abstract-syntax-tree) as a result.  
+    | The execution phase is built upon the AST.
+    
+    | ``Base_statement`` is the base for the all object-nodes participating in the execution phase, it consists of the ``eval()`` method which returns the <value> object.
+    
+    | ``value`` object is handling the known basic-types such as int,string,float,time-stamp
+    | It is able to operate comparison and basic arithmetic operations on mentioned types.
+    
+    | The execution-flow is actually calling the ``eval()`` method on the root-node (per each projection), it goes all the way down, and returns the actual result (``value`` object) from bottom node to root node(all the way up) .
+
+    | **Alias** programming-construct is an essential part of s3-select language, it enables much better programming especially with objects containing many columns or in the case of complex queries.
+    
+    | Upon parsing the statement containing alias construct, it replaces alias with reference to the correct AST-node, on runtime the node is simply evaluated as any other node.
+
+    | There is a risk that self(or cyclic) reference may occur causing stack-overflow(endless-loop), for that concern upon evaluating an alias, it is validated for cyclic reference.
+    
+    | Alias also maintains result-cache, meaning upon using the same alias more than once, it’s not evaluating the same node again(it will return the same result),instead it uses the result from cache.
+
+    | Of Course, per each new row the cache is invalidated.
+        
+
+S3 select parser definition
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    | The implementation of s3-select uses the `boost::spirit <https://www.boost.org/doc/libs/1_71_0/libs/spirit/classic/doc/grammar.html>`_ the definition of s3-select command is according to AWS.
+     
+    | Upon parsing is initiated on input text, and a specific rule is identified, an action which is bound to that rule is executed.
+    | Those actions are building the AST, each action is unique (as its rule), at the end of the process it forms a structure similar to a tree. 
+    
+    | As mentioned, running eval() on the root node, execute the s3-select statement (per projection).
+    | The input stream is accessible to the execution tree, by the scratch-area object, that object is constantly updated per each new row. 
+
+Basic functionalities
+~~~~~~~~~~~~~~~~~~~~~
+
+    | **S3select** has a definite set of functionalities that should be implemented (if we wish to stay compliant with AWS), currently only a portion of it is implemented.
+    
+    | The implemented software architecture supports basic arithmetic expressions, logical and compare expressions, including nested function calls and casting operators, that alone enables the user reasonable flexibility. 
+    | review the bellow feature-table_.
+
+
+
+Memory handling
```

Comment:
could be removed or moved to design doc

Reply:
we should discuss (offline) , what should appear in this doc and what on other docs/repos.

the purpose of this section, is about resource management, and the potential impact on hosting application (RGW/OSD).
very important.

Reply:
you can give some guidelines to the user regarding the impact of memory usage of this feature, but no need to provide that via implementation and design details.
this is usually does by testing and evaluating the memory and perf impact of the feature.
as this usually requires significant effort, would suggest doing that in a later phase